UI feedback of state machine

SWvheerden opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Adds a channel to the base_node parser to listen for state machine publisher to publish its state and info. 
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This PR provides feedback for the user to see what the status and info from the state machine is, like how syncing is, is it syncing etc. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

sdbondi commented on applications/tari_base_node/src/parser.rs:
```diff
@@ -329,6 +336,9 @@ impl Parser {
                 println!("{}", joined);
             },
+            GetStateInfo => {
+                println!("Prints out the status of the base_node state machine");
```
nit:
```suggestion
                println!("Prints out the status of the base node state machine");
```

sdbondi commented on applications/tari_base_node/src/parser.rs:
```diff
@@ -433,6 +443,21 @@ impl Parser {
+        self.executor.spawn(async move {
+            match channel.next().await {
+                None => {
+                    println!("Something went wrong");
```
`None` happens when the channel closes (e.g. if the node shuts down). I wouldn't print any output in this case as it might give the user an impression of instability. The log below could be an `info!` saying that the state machine state channel closed.

sdbondi commented on base_layer/core/src/base_node/states/listening.rs:
```diff
@@ -64,6 +64,7 @@ impl ListeningData {
     pub async fn next_event<B: BlockchainBackend>(&mut self, shared: &mut BaseNodeStateMachine<B>) -> StateEvent {
         info!(target: LOG_TARGET, "Listening for chain metadata updates");
         shared.info = StatusInfo::Listening(ListeningInfo::new());
+        let _ = shared.status_event_publisher.send(shared.info.clone()).await;
```
Ideally, publishing an event would just be wrapped in a function
```suggestion
        shared.publish_status_event(shared.info.clone()).await;
```

sdbondi commented on base_layer/core/src/base_node/states/block_sync.rs:
```diff
@@ -161,6 +161,7 @@ impl BlockSyncStrategy {
     ) -> StateEvent
     {
         shared.info = StatusInfo::BlockSync(BlockSyncInfo::new(None, None, None));
+        let _ = shared.status_event_publisher.send(shared.info.clone()).await;
```
Write and use a `publish_status_event` function

sdbondi reviewed (commented):
LGTM - few small changes

sdbondi reviewed (dismissed):
LGTM - we might want to implement a `set_state_machine_info` function that sets the info and publishes the event to remove the possibility of bugs from not publishing an event. But not NB for now

sdbondi commented on applications/tari_base_node/src/parser.rs:
```diff
@@ -447,6 +457,23 @@ impl Parser {
+    fn process_state_info(&mut self) {
+        let mut channel = self.state_machine_info.clone();
+        self.executor.spawn(async move {
+            match channel.next().await {
```
This would not necessarily have the current state of the base node, or any state change queued up - which would make the command look as if it doesn't do anything - what we really need is a handle function that can query the base node status.

SWvheerden replied:
I covered all the normal states in the channel It should have the state. 
I think the only states not covered, is startup and closing.

sdbondi replied:
Thanks for the comments - yup I see how it works now :)

SWvheerden committed:
Adds a channel to the base_node parser to listen for state machine feedback

Changed listing state display to be more clear
Co-authored-by: Stan Bondi <sdbondi@users.noreply.github.com>
Add comment to describe why the channel will never be empty
